Core/Core/Views/ScannerViewController.swift

```diff
@@ -0,0 +1,217 @@
+            guide.centerYAnchor.constraint(equalTo: view.safeAreaLayoutGuide.centerYAnchor),
+        ])
+
+        let promptContainer = UIView()
+        promptContainer.translatesAutoresizingMaskIntoConstraints = false
+        promptContainer.backgroundColor = UIColor.named(.backgroundDarkest).withAlphaComponent(0.9)
+        promptContainer.layer.cornerRadius = 8
+        view.addSubview(promptContainer)
+        NSLayoutConstraint.activate([
+            promptContainer.centerXAnchor.constraint(equalTo: view.safeAreaLayoutGuide.centerXAnchor),
+            promptContainer.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 80),
+        ])
+
+        let prompt = UILabel()
+        prompt.text = NSLocalizedString("Find a code to scan", bundle: .core, comment: "")
+        prompt.font = .scaledNamedFont(.semibold16)
+        prompt.textColor = .named(.textLightest)
+        promptContainer.addSubview(prompt)
+        prompt.pin(inside: promptContainer, leading: 16, trailing: 16, top: 16, bottom: 16)
+
+        let cancel = UIButton()
+        cancel.translatesAutoresizingMaskIntoConstraints = false
+        cancel.backgroundColor = UIColor.named(.backgroundDarkest).withAlphaComponent(0.9)
+        cancel.setImage(UIImage.icon(.x), for: .normal)
+        cancel.tintColor = .named(.textLightest)
+        cancel.addTarget(self, action: #selector(cancelTapped(_:)), for: .primaryActionTriggered)
+        cancel.layer.cornerRadius = 25
+        
+        view.addSubview(cancel)
+        NSLayoutConstraint.activate([
+            cancel.widthAnchor.constraint(equalToConstant: 50),
+            cancel.heightAnchor.constraint(equalToConstant: 50),
+            cancel.centerXAnchor.constraint(equalTo: view.safeAreaLayoutGuide.centerXAnchor),
+            cancel.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -80),
+        ])
+
+        captureSession.startRunning()
+    }
+
+    func failed() {
+        let alert = UIAlertController(
+            title: NSLocalizedString("Scanning not supported", bundle: .core, comment: ""),
+            message: nil,
+            preferredStyle: .alert
+        )
+        alert.addAction(UIAlertAction(title: NSLocalizedString("OK", bundle: .core, comment: ""), style: .default))
+        present(alert, animated: true)
+        captureSession = nil
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        if (captureSession?.isRunning == false) {
+            captureSession.startRunning()
+        }
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+
+        if (captureSession?.isRunning == true) {
+            captureSession.stopRunning()
+        }
+    }
+
+    func metadataOutput(_ output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
+        captureSession.stopRunning()
```

Comment:
do we want to stopRunning before we've been successful? There are many cases in this function that would leave this as just a black screen, right?

Reply:
You are right, I think only doing this on success would be best. it doesn't make the screen black but it does freeze the frame.